src/java.base/share/classes/java/util/DualPivotQuicksort.java

```diff
@@ -37,16 +37,16 @@
  *
  * There are also additional algorithms, invoked from the Dual-Pivot
  * Quicksort, such as mixed insertion sort, merging of runs and heap
- * sort, counting sort and parallel merge sort.
+ * sort, radix sort, counting sort and parallel merge sort.
  *
  * @author Vladimir Yaroslavskiy
  * @author Jon Bentley
  * @author Josh Bloch
  * @author Doug Lea
  *
- * @version 2018.08.18
+ * @version 2020.06.14
```

Comment:
Vladimir, I would update to 2021.05.06 (+your hash)

Reply:
Laurent, the date in this class is not the date of our last commit,
this date is the date when I have final idea regarding to Radix sort,
therefore, I prefer to keep 2020.06.14

Reply:
as you want, but you should maybe indicate which version corresponds to your master code too.

It would help tracking changes between forks (iarosalvskiy/sorting master)

Reply:
Hi @iaroslavski I'm unconvinced that this work was from 14/06/2020 - I believe this work derives from an unsigned radix sort I implemented on 10/04/2021 https://github.com/richardstartin/radix-sort-benchmark/commit/ab4da230e1d0ac68e5ee2cee38d71c7e7d50f49b#diff-6c13d3fb74f38906677dbfa1a70a123c8e5baf4a39219c81ef121e078d0013bcR226 which has numerous structural similarities to this work:
* Producing all four histograms in one pass
* Skipping passes based on detecting the total in the histogram
* Bailing out of the skip detection if a nonzero value not equal to the total is encountered
* Manually unrolling the LSD radix sort loop in order to avoid array copies

My implementation from 10th April is below for reference:
```java
  public static void unrollOnePassHistogramsSkipLevels(int[] data) {
    int[] histogram1 = new int[257];
    int[] histogram2 = new int[257];
    int[] histogram3 = new int[257];
    int[] histogram4 = new int[257];

    for (int value : data) {
      ++histogram1[(value & 0xFF) + 1];
      ++histogram2[((value >>> 8) & 0xFF) + 1];
      ++histogram3[((value >>> 16) & 0xFF) + 1];
      ++histogram4[(value >>> 24) + 1];
    }
    boolean skipLevel1 = canSkipLevel(histogram1, data.length);
    boolean skipLevel2 = canSkipLevel(histogram2, data.length);
    boolean skipLevel3 = canSkipLevel(histogram3, data.length);
    boolean skipLevel4 = canSkipLevel(histogram4, data.length);

    if (skipLevel1 && skipLevel2 && skipLevel3 && skipLevel4) {
      return;
    }
    int[] copy = new int[data.length];

    int[] source = data;
    int[] dest = copy;

    if (!skipLevel1) {
      for (int i = 1; i < histogram1.length; ++i) {
        histogram1[i] += histogram1[i - 1];
      }
      for (int value : source) {
        dest[histogram1[value & 0xFF]++] = value;
      }
      if (!skipLevel2 || !skipLevel3 || !skipLevel4) {
        int[] tmp = dest;
        dest = source;
        source = tmp;
      }
    }

    if (!skipLevel2) {
      for (int i = 1; i < histogram2.length; ++i) {
        histogram2[i] += histogram2[i - 1];
      }
      for (int value : source) {
        dest[histogram2[(value >>> 8) & 0xFF]++] = value;
      }
      if (!skipLevel3 || !skipLevel4) {
        int[] tmp = dest;
        dest = source;
        source = tmp;
      }
    }

    if (!skipLevel3) {
      for (int i = 1; i < histogram3.length; ++i) {
        histogram3[i] += histogram3[i - 1];
      }
      for (int value : data) {
        dest[histogram3[(value >>> 16) & 0xFF]++] = value;
      }
      if (!skipLevel4) {
        int[] tmp = dest;
        dest = source;
        source = tmp;
      }
    }

    if (!skipLevel4) {
      for (int i = 1; i < histogram4.length; ++i) {
        histogram4[i] += histogram4[i - 1];
      }
      for (int value : source) {
        dest[histogram4[value >>> 24]++] = value;
      }
    }
    if (dest != data) {
      System.arraycopy(dest, 0, data, 0, data.length);
    }
  }

  private static boolean canSkipLevel(int[] histogram, int dataSize) {
    for (int count : histogram) {
      if (count == dataSize) {
        return true;
      } else if (count > 0) {
        return false;
      }
    }
    return true;
  }
```

Moreover, @bourgesl forked my repository on 11/04/2021 and communicated with me about doing so. On 25/04/2021 there was a new implementation of `DualPivotQuicksort` with a signed radix sort but the same structural similarities, and with the same method and variable names in places https://github.com/bourgesl/radix-sort-benchmark/commit/90ff7e427da0fa49f374bff0241fb2487bd87bde#diff-397ce8fd791e2ce508cf9127201bc9ab46264cd2a79fd0487a63569f2e4b59b2R607-R609

```java
    // TODO add javadoc
    private static void radixSort(Sorter sorter, int[] a, int low, int high) {
        int[] b;
        // LBO: prealloc (high - low) +1 element:
        if (sorter == null || (b = sorter.b) == null || b.length < (high - low)) {
            // System.out.println("alloc b: " + (high - low));
            b = new int[high - low];
        }

        int[] count1, count2, count3, count4;
        if (sorter != null) {
            sorter.resetRadixBuffers();
            count1 = sorter.count1;
            count2 = sorter.count2;
            count3 = sorter.count3;
            count4 = sorter.count4;
        } else {
            // System.out.println("alloc radix buffers(4x256)");
            count1 = new int[256];
            count2 = new int[256];
            count3 = new int[256];
            count4 = new int[256];
        }

        for (int i = low; i < high; ++i) {
            --count1[ a[i]         & 0xFF ];
            --count2[(a[i] >>>  8) & 0xFF ];
            --count3[(a[i] >>> 16) & 0xFF ];
            --count4[(a[i] >>> 24) ^ 0x80 ];
        }

        boolean skipLevel4 = canSkipLevel(count4, low - high);
        boolean skipLevel3 = skipLevel4 && canSkipLevel(count3, low - high);
        boolean skipLevel2 = skipLevel3 && canSkipLevel(count2, low - high);

        count1[255] += high;
        count2[255] += high;
        count3[255] += high;
        count4[255] += high;

        // 1 todo process LSD
        for (int i = 255; i > 0; --i) {
            count1[i - 1] += count1[i];
        }

        for (int i = low; i < high; ++i) {
            b[count1[a[i] & 0xFF]++ - low] = a[i];
        }

        if (skipLevel2) {
            System.arraycopy(b, 0, a, low, high - low);
            return;
        }

        // 2
        for (int i = 255; i > 0; --i) {
            count2[i - 1] += count2[i];
        }

        //for (int value : b) {
        //    a[count2[(value >> 8) & 0xFF]++] = value;
        for (int i = low; i < high; ++i) {
            a[count2[(b[i] >> 8) & 0xFF]++] = b[i];
        }

        if (skipLevel3) {
            return;
        }

        // 3
        for (int i = 255; i > 0; --i) {
            count3[i - 1] += count3[i];
        }

        for (int i = low; i < high; ++i) {
            b[count3[(a[i] >> 16) & 0xFF]++ - low] = a[i];
        }

        if (skipLevel4) {
            System.arraycopy(b, 0, a, low, high - low);
            return;
        }

        // 4
        for (int i = 255; i > 0; --i) {
            count4[i - 1] += count4[i];
        }

        // for (int value : b) {
        //    a[count4[ (value >>> 24) ^ 0x80]++] = value;
        for (int i = low; i < high; ++i) {
            a[count4[ (b[i] >>> 24) ^ 0x80]++] = b[i];
        }
    }

    // TODO: add javadoc
    private static boolean canSkipLevel(int[] count, int total) {
        for (int c : count) {
            if (c == 0) {
                continue;
            }
            return c == total;
        }
        return true;
    }
```

Later, the name of the method `canSkipLevel` changed to `skipByte`: https://github.com/bourgesl/radix-sort-benchmark/commit/a693b26b2e2c14cfeedf9c753c9d643096b0e38d#diff-397ce8fd791e2ce508cf9127201bc9ab46264cd2a79fd0487a63569f2e4b59b2R719 - this is the name of the method in the version of this sort you committed on 05/01/2021 https://github.com/richardstartin/sorting/commit/f076073b8b819a9687613903a164e3ed71821769#diff-4b4d68fc834c2ad12a9fb9d316a812221af7c398338ed2ee907d0a795e7aadafR601-R604

```java
    // TODO add javadoc
//  private 
    static void radixSort(Sorter sorter, int[] a, int low, int high) {
//System.out.println("                        Radix !!!");
        int[] count1 = new int[256];
        int[] count2 = new int[256];
        int[] count3 = new int[256];
        int[] count4 = new int[256];

        for (int i = low; i < high; ++i) {
            count1[ a[i]         & 0xFF ]--;
            count2[(a[i] >>>  8) & 0xFF ]--;
            count3[(a[i] >>> 16) & 0xFF ]--;
            count4[(a[i] >>> 24) ^ 0x80 ]--;
        }
        boolean skipByte4 = skipByte(count4, low - high, high, true);
        boolean skipByte3 = skipByte(count3, low - high, high, skipByte4);
        boolean skipByte2 = skipByte(count2, low - high, high, skipByte3);
        boolean skipByte1 = skipByte(count1, low - high, high, skipByte2);

        if (skipByte1) {
//Main.check(a, low, high - 1); // todo
            return;
        }
//        int xorA = Main.getXor(a, low, high);

        int[] b; int offset = low;

        if (sorter == null || (b = (int[]) sorter.b) == null) {
            b = new int[high - low];
        } else {
            offset = sorter.offset;
//System.out.println("      !!!! offset: " + offset);
        }
        int start = low - offset;
        int last = high - offset;


        // 1 todo process LSD
        for (int i = low; i < high; ++i) {
            b[count1[a[i] & 0xFF]++ - offset] = a[i];
        }

//        if (xorA != Main.getXor(a, low, high)) System.out.println("6K_4 1 xor changed");

        if (skipByte2) {
            System.arraycopy(b, start, a, low, high - low);
//Main.check(a, low, high - 1); // todo
            return;
        }

        // 2
        for (int i = start; i < last; ++i) {
            a[count2[(b[i] >> 8) & 0xFF]++] = b[i];
        }

//        if (xorA != Main.getXor(a, low, high)) System.out.println("6K_4 2 xor changed");

        if (skipByte3) {
//Main.check(a, low, high - 1); // todo
            return;
        }

        // 3
        for (int i = low; i < high; ++i) {
            b[count3[(a[i] >> 16) & 0xFF]++ - offset] = a[i];
        }

//        if (xorA != Main.getXor(a, low, high)) System.out.println("6K_4 3 xor changed");

        if (skipByte4) {
            System.arraycopy(b, start, a, low, high - low);
//Main.check(a, low, high - 1); // todo
            return;
        }

//        if (xorA != Main.getXor(a, low, high)) System.out.println("6K_4 4 xor changed");
        // 4
        for (int i = start; i < last; ++i) {
            a[count4[(b[i] >>> 24) ^ 0x80]++] = b[i];
        }
//        if (xorA != Main.getXor(a, low, high)) System.out.println("6K_4 5 xor changed");
//Main.check(a, low, high - 1); // todo
    }

    // TODO: add javadoc
    private static boolean skipByte(int[] count, int total, int high, boolean prevSkip) {
        if (prevSkip) {
            for (int c : count) {
                if (c == 0) {
                    continue;
                }
                if (c == total) {
                    return true;
                }
                break;
            }
        }
        // todo create historgam
        count[255] += high;

        for (int i = 255; i > 0; --i) {
            count[i - 1] += count[i];
        }
        return false;
    }
```

`skipByte` was later renamed to `passLevel` here, which is the name used in this PR: https://github.com/richardstartin/sorting/commit/22357e407d3ae7a1e159e06fe4afbb2c57f7d34c#diff-4b4d68fc834c2ad12a9fb9d316a812221af7c398338ed2ee907d0a795e7aadaf

Given the structural similarities mentioned, the chronological order of these commits, and the demonstrable provenance of the method name `passLevel` from `canSkipLevel` and since you have patented sort algorithms in the past, I want to make sure that it is recognised that this work is derived from my own.

Reply:
You misunderstood my approach:
- vladimir & tagir discussed radix sorts since previous work on DPQS in 2019
- I enjoyed reading your blog post testing the performance of your radix sort vs Arrays.sort()
- I tested and forked your radix-sort-benchmark to reproduce your experiments on openjdk16 (dpqs 19.11)
- Vladimir proposed his own radixsort()
- I did port DPQS impls in my fork of your benchmark to make a global comparison: dpqs 11, 19, 21 vs yours + arrays.sort(): it helped comparing implementations and decide when radix sort wins depending on dataset presortness
- I tried many variants on my github repositories, but Vladimir never merged any of my change as he is not a regular github user (clone, fork, merge).

Our goal is not to underestimate your work (sort + benchmark) but Vladimir wrote his own code, me many experiments (tests, benchmarks) to obtain this original patch, written by Vladimir, with his radix sort implementation working on any int/long/float/double arrays, following the GPLv2 license.

You gave me motivation to make Arrays.sort() faster and we worked hard to write, test and benchmark this patch to be ready for OpenJDK 17

Reply:
@iaroslavski in light of the above, would you please clarify: 
* where you got the code in commit https://github.com/richardstartin/sorting/commit/f076073b8b819a9687613903a164e3ed71821769#diff-4b4d68fc834c2ad12a9fb9d316a812221af7c398338ed2ee907d0a795e7aadafR601-R604 from
* that you wrote it yourself and you did not get it from @bourgesl (whose work was demonstrably derived from my own)
* if you did not get it from @bourgesl, how did you independently start with the same variable and method names as found in his implementation?

Reply:
Perhaps we can resolve this issue in private - my email address is on my profile (or in the commits in `radix-sort-benchmark`)?

Reply:
@richardstartin The ideas to take 4 histograms at once or use check to skip digits are very simple and come not only from you, for example, see article from 2001, http://stereopsis.com/radix.html

When Laurent and I discussed our Radix sort, your code was demonstrated. I wrote my implementation and at the first version I took the same names (even not perfect for my cases) and later I renamed them to better values.

You can see that from functional point of view your code and my code are different, they process skipped digits in different manner, No your code was copied.

Date 2020.06.14 means the start of my activities on Radix sort, not final version.

Let me know, if you have any questions

Reply:
@richardstartin And one more addon: my first version of Radix sort, see my github https://github.com/iaroslavski/sorting/tree/master/radixsort uses another name, like skipBytes, then renamed to passLevel.
So, the common part is "skip". And this method has different number of parameters. I don't see any collision with your code.

Reply:
@iaroslavski I would prefer to discuss this in private than here, but my argument is that the name `skipByte` came from Laurent's code, and that Laurent's code was clearly derived from my own within a fork of my repository. I linked the commits where you changed `skipByte` to `passLevel` and Laurent changed my name `canSkipLevel` to `skipByte`. 

For me, this raises questions about the independence of your work from Laurent's, and Laurent's work is clearly derived from my own (and I don't think anyone is disputing the latter). I would be happy to sort this out in private.

Reply:
In private correspondence with Vladimir, it was explained that where Vladimir's code and Laurent's code are identical, including typos ([Vladimir's code](https://github.com/iaroslavski/sorting/commit/f076073b8b819a9687613903a164e3ed71821769#diff-4b4d68fc834c2ad12a9fb9d316a812221af7c398338ed2ee907d0a795e7aadafR672), [Laurent's code](https://github.com/bourgesl/radix-sort-benchmark/commit/a693b26b2e2c14cfeedf9c753c9d643096b0e38d#diff-397ce8fd791e2ce508cf9127201bc9ab46264cd2a79fd0487a63569f2e4b59b2R719)) it is because Vladimir sent the code to Laurent, not the other way around, therefore Vladimir's code does not derive from Laurent's, and it does not derive from mine. I can only trust that this is the case, so please disregard my claim that this is derivative work when reviewing this PR.

Reply:
For what it's worth, I benchmarked this implementation radix sort ([adaptated to here to fit in to my harness](https://github.com/richardstartin/radix-sort-benchmark/commit/07169e8e8602152cfda859baa159db165bf5fcab#diff-6c13d3fb74f38906677dbfa1a70a123c8e5baf4a39219c81ef121e078d0013bcR681-R710782)) against a [signed variant](https://github.com/richardstartin/radix-sort-benchmark/commit/07169e8e8602152cfda859baa159db165bf5fcab#diff-6c13d3fb74f38906677dbfa1a70a123c8e5baf4a39219c81ef121e078d0013bcR396-R478) of what I have claimed this work was derived from and the proposed implementation does not perform favourably on uniformly random data:


```
Benchmark                                         (bits)  (padding)  (scenario)  (seed)   (size)  Mode  Cnt      Score     Error  Units
RadixSortBenchmark.jdk                                17          7     UNIFORM       0  1000000  avgt    5  11301.950 ± 113.691  us/op
RadixSortBenchmark.jdk                                23          7     UNIFORM       0  1000000  avgt    5  11792.351 ±  60.757  us/op
RadixSortBenchmark.jdk                                30          7     UNIFORM       0  1000000  avgt    5  11184.616 ±  67.094  us/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned      17          7     UNIFORM       0  1000000  avgt    5   9564.626 ±  69.497  us/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned      23          7     UNIFORM       0  1000000  avgt    5   9432.085 ±  58.983  us/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned      30          7     UNIFORM       0  1000000  avgt    5  10772.975 ±  51.848  us/op

```

I believe the root cause is a defect in the mechanism employed to skip passes as can be seen by the increased number of instructions and cycles here. In the proposed implementation, instructions is roughly constant as a function of bits. In the case where all passes must be performed (bits = 30), IPC is superior in `unrollOnePassHistogramsSkipLevelsSigned`.

```
Benchmark                                                               (bits)  (padding)  (scenario)  (seed)   (size)  Mode  Cnt         Score     Error      Units
RadixSortBenchmark.jdk:cycles                                               17          7     UNIFORM       0  1000000  avgt       34976971.877                 #/op
RadixSortBenchmark.jdk:instructions                                         17          7     UNIFORM       0  1000000  avgt       70121142.003                 #/op
RadixSortBenchmark.jdk:cycles                                               23          7     UNIFORM       0  1000000  avgt       32369970.385                 #/op
RadixSortBenchmark.jdk:instructions                                         23          7     UNIFORM       0  1000000  avgt       70201664.963                 #/op
RadixSortBenchmark.jdk:cycles                                               30          7     UNIFORM       0  1000000  avgt       30789736.602                 #/op
RadixSortBenchmark.jdk:instructions                                         30          7     UNIFORM       0  1000000  avgt       70180942.122                 #/op
RadixSortBenchmark.jdk:IPC                                                  30          7     UNIFORM       0  1000000  avgt              2.279            insns/clk
RadixSortBenchmark.unrollOnePassSkipLevelsSigned:cycles                     17          7     UNIFORM       0  1000000  avgt       26983994.479                 #/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned:instructions               17          7     UNIFORM       0  1000000  avgt       62065304.827                 #/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned:cycles                     23          7     UNIFORM       0  1000000  avgt       26161703.124                 #/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned:instructions               23          7     UNIFORM       0  1000000  avgt       63102683.167                 #/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned:cycles                     30          7     UNIFORM       0  1000000  avgt       29780103.795                 #/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned:instructions               30          7     UNIFORM       0  1000000  avgt       74437097.025                 #/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned:IPC                        30          7     UNIFORM       0  1000000  avgt              2.500            insns/clk

```

The biggest difference in executed code appears to be that bounds checks are not eliminated in the first counting pass in the proposed implementation:

```
....[Hottest Region 1]..............................................................................
c2, level 4, io.github.richardstartin.radixsort.RadixSort::jdk, version 402 (166 bytes) 

            0x00007f0d2fb20579: cmp    $0x100,%r10d
            0x00007f0d2fb20580: jae    0x00007f0d2fb21344
            0x00007f0d2fb20586: decl   0x10(%rdx,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::jdk@107 (line 697)
            0x00007f0d2fb2058b: cmp    $0x1,%esi
            0x00007f0d2fb2058e: jle    0x00007f0d2fb2134c
            0x00007f0d2fb20594: mov    $0x1,%r11d
            0x00007f0d2fb2059a: xor    %ecx,%ecx
            0x00007f0d2fb2059c: nopl   0x0(%rax)          ;*aload_2 {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::jdk@41 (line 694)
  0.06%  ↗  0x00007f0d2fb205a0: movzbl 0x10(%rbp,%r11,4),%r10d  ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@49 (line 694)
  0.18%  │  0x00007f0d2fb205a6: decl   0x10(%r8,%r10,4)   ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@54 (line 694)
  1.37%  │  0x00007f0d2fb205ab: mov    0x10(%rbp,%r11,4),%r10d
  0.42%  │  0x00007f0d2fb205b0: shr    $0x8,%r10d
  0.34%  │  0x00007f0d2fb205b4: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@66 (line 695)
  0.50%  │  0x00007f0d2fb205b8: decl   0x10(%r9,%r10,4)   ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@71 (line 695)
  1.75%  │  0x00007f0d2fb205bd: mov    0x10(%rbp,%r11,4),%r10d
  0.42%  │  0x00007f0d2fb205c2: shr    $0x10,%r10d
  0.12%  │  0x00007f0d2fb205c6: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@84 (line 696)
  0.36%  │  0x00007f0d2fb205ca: decl   0x10(%rbx,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@89 (line 696)
  1.73%  │  0x00007f0d2fb205cf: mov    0x10(%rbp,%r11,4),%r10d
  0.26%  │  0x00007f0d2fb205d4: shr    $0x18,%r10d
  0.38%  │  0x00007f0d2fb205d8: xor    $0x80,%r10d        ;*ixor {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@102 (line 697)
  0.28%  │  0x00007f0d2fb205df: cmp    $0x100,%r10d
         │  0x00007f0d2fb205e6: jae    0x00007f0d2fb20bab
  0.38%  │  0x00007f0d2fb205ec: decl   0x10(%rdx,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@107 (line 697)
  1.25%  │  0x00007f0d2fb205f1: movzbl 0x14(%rbp,%r11,4),%r10d  ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@49 (line 694)
  1.63%  │  0x00007f0d2fb205f7: decl   0x10(%r8,%r10,4)   ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@54 (line 694)
  1.71%  │  0x00007f0d2fb205fc: mov    0x14(%rbp,%r11,4),%r10d
  0.16%  │  0x00007f0d2fb20601: shr    $0x8,%r10d
  0.18%  │  0x00007f0d2fb20605: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@66 (line 695)
  0.36%  │  0x00007f0d2fb20609: decl   0x10(%r9,%r10,4)   ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@71 (line 695)
  2.05%  │  0x00007f0d2fb2060e: mov    0x14(%rbp,%r11,4),%r10d
  0.14%  │  0x00007f0d2fb20613: shr    $0x10,%r10d
  0.38%  │  0x00007f0d2fb20617: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@84 (line 696)
  0.14%  │  0x00007f0d2fb2061b: decl   0x10(%rbx,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@89 (line 696)
  1.51%  │  0x00007f0d2fb20620: mov    0x14(%rbp,%r11,4),%r10d
  0.14%  │  0x00007f0d2fb20625: shr    $0x18,%r10d
  0.32%  │  0x00007f0d2fb20629: xor    $0x80,%r10d        ;*ixor {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@102 (line 697)
  0.12%  │  0x00007f0d2fb20630: cmp    $0x100,%r10d
         │  0x00007f0d2fb20637: jae    0x00007f0d2fb20ba8
  0.78%  │  0x00007f0d2fb2063d: decl   0x10(%rdx,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@107 (line 697)
  1.04%  │  0x00007f0d2fb20642: add    $0x2,%r11d         ;*iinc {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::jdk@108 (line 693)
  0.64%  │  0x00007f0d2fb20646: cmp    %esi,%r11d
         ╰  0x00007f0d2fb20649: jl     0x00007f0d2fb205a0  ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::jdk@38 (line 693)
            0x00007f0d2fb2064f: cmp    %r13d,%r11d
            0x00007f0d2fb20652: jge    0x00007f0d2fb206ad  ;*aload_2 {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::jdk@41 (line 694)
            0x00007f0d2fb20654: movzbl 0x10(%rbp,%r11,4),%r10d  ;*iand {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::jdk@49 (line 694)
            0x00007f0d2fb2065a: decl   0x10(%r8,%r10,4)   ;*iastore {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::jdk@54 (line 694)
....................................................................................................
 21.11%  <total for region 1>

```

Whereas they have been eliminated in the first pass of `unrollOnePassHistogramsSkipLevelsSigned`:

```
....[Hottest Region 1]..............................................................................
c2, level 4, io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned, version 402 (544 bytes) 

            0x00007f123035bb6c: incl   0x14(%rbp,%rcx,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@106 (line 405)
            0x00007f123035bb70: incl   0x14(%r9,%r10,4)   ;*iastore {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@123 (line 406)
            0x00007f123035bb75: cmp    $0x1,%esi
            0x00007f123035bb78: jle    0x00007f123035cc28
            0x00007f123035bb7e: mov    $0x1,%edi
            0x00007f123035bb83: mov    $0x1,%ebx
            0x00007f123035bb88: nopl   0x0(%rax,%rax,1)   ;*aload {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@43 (line 402)
  0.12%  ↗  0x00007f123035bb90: mov    $0x80000000,%r10d
  0.06%  │  0x00007f123035bb96: add    0x10(%rdx,%rdi,4),%r10d  ;*isub {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::signed@3 (line 476)
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@53 (line 403)
  0.04%  │  0x00007f123035bb9b: mov    %r10d,%r8d
  0.10%  │  0x00007f123035bb9e: shr    $0x18,%r8d         ;*iushr {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@116 (line 406)
  0.12%  │  0x00007f123035bba2: mov    %r10d,%r11d
  0.08%  │  0x00007f123035bba5: shr    $0x10,%r11d
  0.08%  │  0x00007f123035bba9: movzbl %r11b,%r11d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@99 (line 405)
  0.08%  │  0x00007f123035bbad: mov    %r10d,%ecx
  0.04%  │  0x00007f123035bbb0: shr    $0x8,%ecx
  0.08%  │  0x00007f123035bbb3: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@59 (line 403)
  0.14%  │  0x00007f123035bbb7: mov    0x18(%rsp),%r14
  0.16%  │  0x00007f123035bbbc: incl   0x14(%r14,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@66 (line 403)
  0.36%  │  0x00007f123035bbc1: movzbl %cl,%ecx           ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@79 (line 404)
  0.06%  │  0x00007f123035bbc4: mov    0x10(%rsp),%r10
  0.10%  │  0x00007f123035bbc9: incl   0x14(%r10,%rcx,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@86 (line 404)
  0.26%  │  0x00007f123035bbce: incl   0x14(%rbp,%r11,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@106 (line 405)
  0.28%  │  0x00007f123035bbd3: incl   0x14(%r9,%r8,4)    ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@123 (line 406)
  0.20%  │  0x00007f123035bbd8: mov    $0x80000000,%r10d
  0.10%  │  0x00007f123035bbde: add    0x14(%rdx,%rdi,4),%r10d  ;*isub {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::signed@3 (line 476)
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@53 (line 403)
  0.08%  │  0x00007f123035bbe3: mov    %r10d,%r8d
  0.12%  │  0x00007f123035bbe6: shr    $0x18,%r8d         ;*iushr {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@116 (line 406)
  0.04%  │  0x00007f123035bbea: mov    %r10d,%r11d
  0.06%  │  0x00007f123035bbed: shr    $0x10,%r11d
  0.06%  │  0x00007f123035bbf1: movzbl %r11b,%r11d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@99 (line 405)
  0.14%  │  0x00007f123035bbf5: mov    %r10d,%ecx
  0.04%  │  0x00007f123035bbf8: shr    $0x8,%ecx
  0.04%  │  0x00007f123035bbfb: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@59 (line 403)
  0.08%  │  0x00007f123035bbff: incl   0x14(%r14,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@66 (line 403)
  0.46%  │  0x00007f123035bc04: movzbl %cl,%ecx           ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@79 (line 404)
  0.10%  │  0x00007f123035bc07: mov    0x10(%rsp),%r10
  0.08%  │  0x00007f123035bc0c: incl   0x14(%r10,%rcx,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@86 (line 404)
  0.22%  │  0x00007f123035bc11: incl   0x14(%rbp,%r11,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@106 (line 405)
  0.30%  │  0x00007f123035bc16: incl   0x14(%r9,%r8,4)    ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@123 (line 406)
  0.22%  │  0x00007f123035bc1b: mov    $0x80000000,%r10d
  0.04%  │  0x00007f123035bc21: add    0x18(%rdx,%rdi,4),%r10d  ;*isub {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::signed@3 (line 476)
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@53 (line 403)
  0.20%  │  0x00007f123035bc26: mov    %r10d,%r8d
  0.18%  │  0x00007f123035bc29: shr    $0x18,%r8d         ;*iushr {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@116 (line 406)
  0.12%  │  0x00007f123035bc2d: mov    %r10d,%r11d
  0.06%  │  0x00007f123035bc30: shr    $0x10,%r11d
  0.20%  │  0x00007f123035bc34: movzbl %r11b,%r11d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@99 (line 405)
  0.10%  │  0x00007f123035bc38: mov    %r10d,%ecx
  0.12%  │  0x00007f123035bc3b: shr    $0x8,%ecx
  0.20%  │  0x00007f123035bc3e: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@59 (line 403)
  0.08%  │  0x00007f123035bc42: incl   0x14(%r14,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@66 (line 403)
  0.18%  │  0x00007f123035bc47: movzbl %cl,%ecx           ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@79 (line 404)
  0.08%  │  0x00007f123035bc4a: mov    0x10(%rsp),%r10
  0.10%  │  0x00007f123035bc4f: incl   0x14(%r10,%rcx,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@86 (line 404)
  0.20%  │  0x00007f123035bc54: incl   0x14(%rbp,%r11,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@106 (line 405)
  0.12%  │  0x00007f123035bc59: incl   0x14(%r9,%r8,4)    ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@123 (line 406)
  0.34%  │  0x00007f123035bc5e: mov    $0x80000000,%r10d
  0.04%  │  0x00007f123035bc64: add    0x1c(%rdx,%rdi,4),%r10d  ;*isub {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::signed@3 (line 476)
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@53 (line 403)
  0.91%  │  0x00007f123035bc69: mov    %r10d,%r8d
  0.02%  │  0x00007f123035bc6c: shr    $0x18,%r8d         ;*iushr {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@116 (line 406)
  0.12%  │  0x00007f123035bc70: mov    %r10d,%r11d
  0.06%  │  0x00007f123035bc73: shr    $0x10,%r11d
  0.06%  │  0x00007f123035bc77: movzbl %r11b,%r11d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@99 (line 405)
  0.06%  │  0x00007f123035bc7b: mov    %r10d,%ecx
         │  0x00007f123035bc7e: shr    $0x8,%ecx
  0.10%  │  0x00007f123035bc81: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@59 (line 403)
  0.08%  │  0x00007f123035bc85: incl   0x14(%r14,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@66 (line 403)
  0.16%  │  0x00007f123035bc8a: movzbl %cl,%ecx           ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@79 (line 404)
         │  0x00007f123035bc8d: mov    0x10(%rsp),%r10
  0.14%  │  0x00007f123035bc92: incl   0x14(%r10,%rcx,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@86 (line 404)
  0.32%  │  0x00007f123035bc97: incl   0x14(%rbp,%r11,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@106 (line 405)
  0.32%  │  0x00007f123035bc9c: incl   0x14(%r9,%r8,4)    ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@123 (line 406)
  0.24%  │  0x00007f123035bca1: mov    $0x80000000,%r10d
  0.12%  │  0x00007f123035bca7: add    0x20(%rdx,%rdi,4),%r10d  ;*isub {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::signed@3 (line 476)
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@53 (line 403)
  0.08%  │  0x00007f123035bcac: mov    %r10d,%r8d
  0.08%  │  0x00007f123035bcaf: shr    $0x18,%r8d         ;*iushr {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@116 (line 406)
  0.06%  │  0x00007f123035bcb3: mov    %r10d,%r11d
  0.08%  │  0x00007f123035bcb6: shr    $0x10,%r11d
  0.10%  │  0x00007f123035bcba: movzbl %r11b,%r11d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@99 (line 405)
  0.14%  │  0x00007f123035bcbe: mov    %r10d,%ecx
  0.14%  │  0x00007f123035bcc1: shr    $0x8,%ecx
  0.14%  │  0x00007f123035bcc4: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@59 (line 403)
  0.08%  │  0x00007f123035bcc8: incl   0x14(%r14,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@66 (line 403)
  0.22%  │  0x00007f123035bccd: movzbl %cl,%ecx           ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@79 (line 404)
  0.10%  │  0x00007f123035bcd0: mov    0x10(%rsp),%r10
  0.14%  │  0x00007f123035bcd5: incl   0x14(%r10,%rcx,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@86 (line 404)
  0.24%  │  0x00007f123035bcda: incl   0x14(%rbp,%r11,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@106 (line 405)
  0.20%  │  0x00007f123035bcdf: incl   0x14(%r9,%r8,4)    ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@123 (line 406)
  0.18%  │  0x00007f123035bce4: mov    $0x80000000,%r10d
  0.04%  │  0x00007f123035bcea: add    0x24(%rdx,%rdi,4),%r10d  ;*isub {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::signed@3 (line 476)
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@53 (line 403)
  0.12%  │  0x00007f123035bcef: mov    %r10d,%r8d
  0.04%  │  0x00007f123035bcf2: shr    $0x18,%r8d         ;*iushr {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@116 (line 406)
  0.08%  │  0x00007f123035bcf6: mov    %r10d,%r11d
  0.16%  │  0x00007f123035bcf9: shr    $0x10,%r11d
  0.06%  │  0x00007f123035bcfd: movzbl %r11b,%r11d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@99 (line 405)
  0.06%  │  0x00007f123035bd01: mov    %r10d,%ecx
  0.04%  │  0x00007f123035bd04: shr    $0x8,%ecx
  0.02%  │  0x00007f123035bd07: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@59 (line 403)
  0.12%  │  0x00007f123035bd0b: incl   0x14(%r14,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@66 (line 403)
  0.22%  │  0x00007f123035bd10: movzbl %cl,%ecx           ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@79 (line 404)
  0.06%  │  0x00007f123035bd13: mov    0x10(%rsp),%r10
  0.04%  │  0x00007f123035bd18: incl   0x14(%r10,%rcx,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@86 (line 404)
  0.20%  │  0x00007f123035bd1d: incl   0x14(%rbp,%r11,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@106 (line 405)
  0.28%  │  0x00007f123035bd22: incl   0x14(%r9,%r8,4)    ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@123 (line 406)
  0.46%  │  0x00007f123035bd27: mov    $0x80000000,%r10d
  0.08%  │  0x00007f123035bd2d: add    0x28(%rdx,%rdi,4),%r10d  ;*isub {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::signed@3 (line 476)
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@53 (line 403)
  0.08%  │  0x00007f123035bd32: mov    %r10d,%r8d
  0.04%  │  0x00007f123035bd35: shr    $0x18,%r8d         ;*iushr {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@116 (line 406)
  0.10%  │  0x00007f123035bd39: mov    %r10d,%r11d
  0.04%  │  0x00007f123035bd3c: shr    $0x10,%r11d
  0.06%  │  0x00007f123035bd40: movzbl %r11b,%r11d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@99 (line 405)
  0.04%  │  0x00007f123035bd44: mov    %r10d,%ecx
  0.10%  │  0x00007f123035bd47: shr    $0x8,%ecx
  0.10%  │  0x00007f123035bd4a: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@59 (line 403)
  0.10%  │  0x00007f123035bd4e: incl   0x14(%r14,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@66 (line 403)
  0.28%  │  0x00007f123035bd53: movzbl %cl,%ecx           ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@79 (line 404)
  0.14%  │  0x00007f123035bd56: mov    0x10(%rsp),%r10
  0.10%  │  0x00007f123035bd5b: incl   0x14(%r10,%rcx,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@86 (line 404)
  0.36%  │  0x00007f123035bd60: incl   0x14(%rbp,%r11,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@106 (line 405)
  0.20%  │  0x00007f123035bd65: incl   0x14(%r9,%r8,4)    ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@123 (line 406)
  0.28%  │  0x00007f123035bd6a: mov    $0x80000000,%r10d
  0.16%  │  0x00007f123035bd70: add    0x2c(%rdx,%rdi,4),%r10d  ;*isub {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::signed@3 (line 476)
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@53 (line 403)
  0.08%  │  0x00007f123035bd75: mov    %r10d,%r8d
  0.10%  │  0x00007f123035bd78: shr    $0x18,%r8d         ;*iushr {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@116 (line 406)
  0.08%  │  0x00007f123035bd7c: mov    %r10d,%r11d
  0.08%  │  0x00007f123035bd7f: shr    $0x10,%r11d
  0.14%  │  0x00007f123035bd83: movzbl %r11b,%r11d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@99 (line 405)
  0.02%  │  0x00007f123035bd87: mov    %r10d,%ecx
  0.02%  │  0x00007f123035bd8a: shr    $0x8,%ecx
  0.20%  │  0x00007f123035bd8d: movzbl %r10b,%r10d        ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@59 (line 403)
  0.22%  │  0x00007f123035bd91: incl   0x14(%r14,%r10,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@66 (line 403)
  0.22%  │  0x00007f123035bd96: movzbl %cl,%ecx           ;*iand {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@79 (line 404)
  0.12%  │  0x00007f123035bd99: mov    0x10(%rsp),%r10
  0.04%  │  0x00007f123035bd9e: incl   0x14(%r10,%rcx,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@86 (line 404)
  0.22%  │  0x00007f123035bda3: incl   0x14(%rbp,%r11,4)  ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@106 (line 405)
  0.20%  │  0x00007f123035bda8: incl   0x14(%r9,%r8,4)    ;*iastore {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@123 (line 406)
  0.26%  │  0x00007f123035bdad: add    $0x8,%edi          ;*iinc {reexecute=0 rethrow=0 return_oop=0}
         │                                                ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@124 (line 402)
  0.10%  │  0x00007f123035bdb0: cmp    %esi,%edi
         ╰  0x00007f123035bdb2: jl     0x00007f123035bb90  ;*if_icmpge {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@40 (line 402)
            0x00007f123035bdb8: cmp    %eax,%edi
            0x00007f123035bdba: jge    0x00007f123035be09  ;*aload {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@43 (line 402)
            0x00007f123035bdbc: mov    $0x80000000,%esi
            0x00007f123035bdc1: add    0x10(%rdx,%rdi,4),%esi  ;*isub {reexecute=0 rethrow=0 return_oop=0}
                                                          ; - io.github.richardstartin.radixsort.RadixSort::signed@3 (line 476)
                                                          ; - io.github.richardstartin.radixsort.RadixSort::unrollOnePassHistogramsSkipLevelsSigned@53 (line 403)
....................................................................................................
 18.02%  <total for region 1>

```

So, I think there may be some room for improvement in this submission. Benchmarking is hard, and it's possible I have made a mistake or have misinterpreted the results, but I think such a big difference warrants investigation before pushing this change to so many users. I ran these benchmarks on JDK 11.0.11. I have run a range of randomised tests to verify that the two implementations produce the same results.

Reply:
So the issue of not skipping passes was my fault in the translation process, so not something to worry about, though after [fixing that](https://github.com/richardstartin/radix-sort-benchmark/commit/ccbee984c6a0e0f50c30de59e1a5e9fbcad89510) the original implementation still has the edge because of the bounds checks on the `xor` not getting eliminated.

```
Benchmark                                         (bits)  (padding)  (scenario)  (seed)   (size)  Mode  Cnt      Score    Error  Units
RadixSortBenchmark.jdk                                17          7     UNIFORM       0  1000000  avgt    5  10432.408 ± 87.024  us/op
RadixSortBenchmark.jdk                                23          7     UNIFORM       0  1000000  avgt    5   9465.990 ± 40.598  us/op
RadixSortBenchmark.jdk                                30          7     UNIFORM       0  1000000  avgt    5  11189.146 ± 50.972  us/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned      17          7     UNIFORM       0  1000000  avgt    5   9546.963 ± 41.698  us/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned      23          7     UNIFORM       0  1000000  avgt    5   9412.114 ± 43.081  us/op
RadixSortBenchmark.unrollOnePassSkipLevelsSigned      30          7     UNIFORM       0  1000000  avgt    5  10823.618 ± 64.311  us/op

```

Reply:
Great analysis on C2, richard.

maybe (x ^ 0x80) &0xFF would help C2 to eliminate bound checks...

Reply:
I don't know Laurent, I find the handling of signed order over-complicated. Subtracting `Integer.MIN_VALUE` is really cheap...

Reply:
I made a JMH test on jdk16 to test count4 (xor) performance:
https://github.com/bourgesl/nearly-optimal-mergesort-code/tree/master/sort-bench/results/count_xor


Benchmark (size) Mode Cnt Score Error Units
ArrayXorBenchmark.arrayAndOriginal 1000000 avgt 20 684561,951 ± 2177,120 ns/op
ArrayXorBenchmark.arrayXorOriginal 1000000 avgt 20 777255,465 ± 1815,136 ns/op
ArrayXorBenchmark.arrayXor_Masked 1000000 avgt 20 814163,377 ± 2665,436 ns/op
ArrayXorBenchmark.arrayXor_Unsafe 1000000 avgt 20 707273,922 ± 2994,380 ns/op

Masked xor does not get optimized by c2 too.

Using Unsafe is better, see:
https://github.com/bourgesl/nearly-optimal-mergesort-code/blob/master/sort-bench/src/main/java/edu/sorting/bench/ArrayXorBenchmark.java

If you want, I could make another radixsort() using on or off heap count buffers and Unsafe, as I did in Marlin to avoid bound checks...

Reply:
I think an uncontroversial and efficient solution would be to replace
```
count4[(a[i] >>> 24) ^ 0x80]--;
```
with
```
count4[(a[i] >>> 24) & 0xFF]--;
```
and after finishing counting, first half of count4 should be swapped with second half of count4.

Reply:
FYI I made another radixsortNew() using Unsafe to access to all count arrays:
https://github.com/bourgesl/nearly-optimal-mergesort-code/blob/dbfbd731ffd798defc75528cc1db04063bdb4619/src/edu/sorting/DualPivotQuicksort202105.java#L795

But it is only 2% faster in radix-sort-benchmark (1M arrays):
https://github.com/bourgesl/radix-sort-benchmark/blob/main/results/2021-ref/cmp-16-1M-full.log

It looks not worth the extra complexity for few percents, I think.

Reply:
I will try using VarHandle (17) and see if it can help to make count buffers off-heap and use direct read/write ... maybe 10% faster ? less gc work too.
Maybe auxiliary buffers could be made off-heap if it can remove bound checks too....

Reply:
> Hi @iaroslavski I'm unconvinced that this work was from 14/06/2020 - I believe this work derives from an unsigned radix sort I implemented on 10/04/2021 [richardstartin/radix-sort-benchmark@ab4da23#diff-6c13d3fb74f38906677dbfa1a70a123c8e5baf4a39219c81ef121e078d0013bcR226](https://github.com/richardstartin/radix-sort-benchmark/commit/ab4da230e1d0ac68e5ee2cee38d71c7e7d50f49b#diff-6c13d3fb74f38906677dbfa1a70a123c8e5baf4a39219c81ef121e078d0013bcR226) which has numerous structural similarities to this work:
> Moreover, @bourgesl forked my repository on 11/04/2021 and communicated with me about doing so. On 25/04/2021 there was a new implementation of `DualPivotQuicksort` with a signed radix sort but the same structural similarities, and with the same method and variable names in places [bourgesl/radix-sort-benchmark@90ff7e4#diff-397ce8fd791e2ce508cf9127201bc9ab46264cd2a79fd0487a63569f2e4b59b2R607-R609](https://github.com/bourgesl/radix-sort-benchmark/commit/90ff7e427da0fa49f374bff0241fb2487bd87bde#diff-397ce8fd791e2ce508cf9127201bc9ab46264cd2a79fd0487a63569f2e4b59b2R607-R609)

@iaroslavski The attribution is not clear here. Can you provide a summary as to who is contributing to this patch? I can't tell if all involved have signed the OCA or not. I'm sure there will be questions about space/time trade-offs with radix sort but I think it's important to first establish the origins of this patch first.

Reply:
@AlanBateman There was discussion with Richard Startin, where Laurent and I explained the origin of my patch. There was misunderstanding based on git changes only. Finally Richard wrote comment in PR on May 13, 2021,
I duplicate it here:

"In private correspondence with Vladimir, it was explained that where Vladimir's code and Laurent's code are identical, including typos (Vladimir's code, Laurent's code) it is because Vladimir sent the code to Laurent, not the other way around, therefore Vladimir's code does not derive from Laurent's, and it does not derive from mine. I can only trust that this is the case, so please disregard my claim that this is derivative work when reviewing this PR."

@AlanBateman Vertical pipeline of PR hides comments in the middle and you have to click on "Show more..." to see all comments. There are no claims related to the origin of my patch, it doesn't violate any rights.

Reply:
> @AlanBateman Vertical pipeline of PR hides comments in the middle and you have to click on "Show more..." to see all comments. There are no claims related to the origin of my patch, it doesn't violate any rights.

There is a comment from richardstartin suggesting that something was derived from code in his repo. Is this a benchmark that is not part of this PR? Only asking because I can't find him on OCA signatories. You can use the Skara /contributor command to list the contributors.

Reply:
Mentioned benchmarking is not part of this PR. Our benchmarking was done by Laurent.

Reply:
@AlanBateman my claim was that the implementation was derived from my implementation, and demonstrated a sequence of name changes after @bourgesl forked my repository containing a structurally similar radix sort implementation and benchmarks, in order to provide circumstantial evidence for my claim. Via email @iaroslavski told me that this was not the case, which I decided to accept at face value. So please judge this PR on its merits, and disregard the claims made in these comments. I have not signed an OCA but do not want to block this PR if the space time tradeoff is deemed acceptable.